out_s3: support s3 tagging for uploaded objects

dmytro-sirenko opened the pull request:
This change implements applying S3 tags to uploaded objects. 

It adds a new key to `s3` output configuration, called `s3_object_tag`; its values are key-value pairs of S3 tags to apply to uploaded objects.

A configuration example: 

```ini
[OUTPUT]
    match           *
    name            s3
    region          eu-west-1
    profile         my-aws-profile
    bucket          an-s3-logs-bucket
    s3_key_format   /logs/tmp-dsi-test/$TAG/
    s3_object_tag   key1  value1
    s3_object_tag   key2  value2
```

<!-- Issue number, if available. E.g. "Fixes #31", "Addresses #42, #77" -->

Implements a feature request in https://github.com/fluent/fluent-bit/issues/9685

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:

- [x] Example configuration file for the change
- [x] Debug log output from testing the change
- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found
<!--
Please refer to the Developer Guide for instructions on building Fluent Bit with Valgrind support:
https://github.com/fluent/fluent-bit/blob/master/DEVELOPER_GUIDE.md#valgrind
Invoke Fluent Bit and Valgrind as: $ valgrind --leak-check=full ./bin/fluent-bit <args>
-->

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.

- [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [N/A] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**
<!-- Docs can be edited at https://github.com/fluent/fluent-bit-docs -->
- [x] Documentation required for this feature

<!--  Doc PR (not required but highly recommended) -->

**Backporting**
<!--
PRs targeting the default master branch will go into the next major release usually.
If this PR should be backported to the current or earlier releases then please submit a PR for that particular branch.
-->
- [ ] Backport to latest stable release.

<!--  Other release PR (not required but highly recommended for quick turnaround) -->
----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added S3 object tagging via a new s3_object_tag configuration (key/value pairs). Tags are sent with uploads (single PUT and multipart) as a URL-encoded x-amz-tagging header; the encoded tag string is logged when set.
  * Tagging data is initialized on startup and cleaned up on shutdown to avoid leaks.

* **Tests**
  * Updated S3 tests to cover object tagging, multipart flows, and tags with special characters.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

dmytro-sirenko committed:
Implement object tagging for s3 output.

This adds a new key to the `s3` output configuration, `object_tagging`:

```
[OUTPUT]
    match           *
    name            s3
    region          eu-west-1
    profile         sre
    bucket          anro-test-fluent-bit
    s3_key_format   /logs/tmp-dsi-test/$TAG/
    object_tagging  such=objects&much=tags
```

Signed-off-by: Dmytro Sirenko <dmytro.sirenko@mongodb.com>

dmytro-sirenko committed:
out_s3: test `object_tagging` configuration setting

Signed-off-by: Dmytro Sirenko <dmytro.sirenko@mongodb.com>

dmytro-sirenko commented:
About valgrind: I run 

```sh
$ env TEST_PRESET=valgrind ./run_code_analysis.sh
...
+ ctest --output-on-failure . -DExperimentalMemCheck
...
        Start  55: flb-rt-out_s3
 55/133 Test  #55: flb-rt-out_s3 ....................................   Passed   19.00 sec
```

My changes to tests/runtime/out_s3.c should exercise the code path that touches my changes. 

Are there other valgrind checks I can do?

cosmo0920 reviewed (changes requested):
We need to use `FLB_CONFIG_MAP_SLIST_1` type in this type of configurations.

dmytro-sirenko committed:
out_s3: add `SLIST_1`-based `object_tag` configuration

This adds a more idiomatic way to enumerate key-value pairs in a config:

```ini
[OUTPUT]
  name           s3
  ...
  s3_object_tag  key1  value1
  s3_object_tag  key2  value2
```

Tests:
- two simple key-value pairs
- one simple key-value pair
- zero `s3_object_tag`s
- a value with spaces and special symbols

Signed-off-by: Dmytro Sirenko <dmytro.sirenko@mongodb.com>

dmytro-sirenko committed:
out_s3: clean up url-encoded `object_tagging` parameter

Signed-off-by: Dmytro Sirenko <dmytro.sirenko@mongodb.com>

dmytro-sirenko commented:
I ran valgrind on a live upload with `s3_object_tag` in a devcontainer:

```sh
[2025/11/05 13:49:22.53012711] [ info] [output:s3:s3.1] Successfully uploaded object /logs/tmp-dsi-test/dummy.local/-objectIE4gOr7g
==25862== 
==25862== HEAP SUMMARY:
==25862==     in use at exit: 0 bytes in 0 blocks
==25862==   total heap usage: 20,331 allocs, 20,331 frees, 24,378,904 bytes allocated
==25862== 
==25862== All heap blocks were freed -- no leaks are possible
==25862== 
==25862== For lists of detected and suppressed errors, rerun with: -s
==25862== ERROR SUMMARY: 9 errors from 4 contexts (suppressed: 0 from 0)
```

<details>
<summary>unfold the full log</summary>

```sh
devcontainer$ tail ../tmp/test.conf 

[OUTPUT]
    match       *
    name        s3
    region      eu-west-1
    profile     <redacted>
    bucket      anro-test-fluent-bit
    s3_key_format /logs/tmp-dsi-test/$TAG/
    s3_object_tag such objects
    s3_object_tag much tags

devcontainer$ valgrind --leak-check=full ./bin/fluent-bit -c ../tmp/test.conf
==25862== Memcheck, a memory error detector
==25862== Copyright (C) 2002-2022, and GNU GPL'd, by Julian Seward et al.
==25862== Using Valgrind-3.19.0 and LibVEX; rerun with -h for copyright info
==25862== Command: ./bin/fluent-bit -c ../tmp/test.conf
==25862== 
Fluent Bit v4.2.0
* Copyright (C) 2015-2025 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

______ _                  _    ______ _ _             ___   __  
|  ___| |                | |   | ___ (_) |           /   | /  | 
| |_  | |_   _  ___ _ __ | |_  | |_/ /_| |_  __   __/ /| | `| | 
|  _| | | | | |/ _ \ '_ \| __| | ___ \ | __| \ \ / / /_| |  | | 
| |   | | |_| |  __/ | | | |_  | |_/ / | |_   \ V /\___  |__| |_
\_|   |_|\__,_|\___|_| |_|\__| \____/|_|\__|   \_/     |_(_)___/
...
[0] dummy.local: [[1762350559.249304293, {}], {"message"=>"dummy"}]
[2025/11/05 13:49:20.244494668] [debug] [task] created task=0x579c190 id=0 OK
[2025/11/05 13:49:20.245107585] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2025/11/05 13:49:20.246769835] [debug] [out flush] cb_destroy coro_id=27
[2025/11/05 13:49:20.245240793] [debug] [output:s3:s3.1] task_id=0 assigned to thread #0
[2025/11/05 13:49:20.247774252] [debug] [out flush] cb_destroy coro_id=27
[2025/11/05 13:49:20.248264085] [debug] [task] destroy task=0x579c190 (task_id=0)
;

^C[2025/11/05 13:49:21] [engine] caught signal (SIGINT)
[0] dummy.local: [[1762350560.245561210, {}], {"message"=>"dummy"}]
[2025/11/05 13:49:21.180550127] [debug] [task] created task=0x57b54b0 id=0 OK
[2025/11/05 13:49:21.180642794] [debug] [output:stdout:stdout.0] task_id=0 assigned to thread #0
[2025/11/05 13:49:21.185091419] [debug] [out flush] cb_destroy coro_id=28
[2025/11/05 13:49:21.185475461] [debug] [out flush] cb_destroy coro_id=28
[2025/11/05 13:49:21.180706252] [debug] [output:s3:s3.1] task_id=0 assigned to thread #0
[2025/11/05 13:49:21.181400669] [ warn] [engine] service will shutdown in max 5 seconds
[2025/11/05 13:49:21.181886044] [debug] [engine] task 0 already scheduled to run, not re-scheduling it.
[2025/11/05 13:49:21.182322961] [ info] [engine] pausing all inputs..
[2025/11/05 13:49:21.182965544] [ info] [input] pausing dummy.0
[2025/11/05 13:49:21.185763627] [debug] [task] destroy task=0x57b54b0 (task_id=0)
[2025/11/05 13:49:21.246213669] [ info] [engine] service has stopped (0 pending tasks)
[2025/11/05 13:49:21.246698419] [ info] [input] pausing dummy.0
[2025/11/05 13:49:21.249782877] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2025/11/05 13:49:21.253974169] [ info] [output:stdout:stdout.0] thread worker #0 stopped
[2025/11/05 13:49:21.279133252] [ info] [output:s3:s3.1] thread worker #0 stopping...
[2025/11/05 13:49:21.279891794] [ info] [output:s3:s3.1] initializing worker
[2025/11/05 13:49:21.281973627] [ info] [output:s3:s3.1] Sending all locally buffered data to S3
[2025/11/05 13:49:21.280945544] [ info] [output:s3:s3.1] thread worker #0 stopped
[2025/11/05 13:49:21.948352794] [debug] [upstream] KA connection #30 to s3.eu-west-1.amazonaws.com:443 is connected
[2025/11/05 13:49:21.949904169] [debug] [http_client] not using http_proxy for header
[2025/11/05 13:49:21.954807003] [debug] [aws_credentials] Requesting credentials from the env provider..
[2025/11/05 13:49:22.51510836] [debug] [upstream] KA connection #30 to s3.eu-west-1.amazonaws.com:443 is now available
[2025/11/05 13:49:22.52457211] [debug] [output:s3:s3.1] PutObject http status=200
[2025/11/05 13:49:22.53012711] [ info] [output:s3:s3.1] Successfully uploaded object /logs/tmp-dsi-test/dummy.local/-objectIE4gOr7g
==25862== 
==25862== HEAP SUMMARY:
==25862==     in use at exit: 0 bytes in 0 blocks
==25862==   total heap usage: 20,331 allocs, 20,331 frees, 24,378,904 bytes allocated
==25862== 
==25862== All heap blocks were freed -- no leaks are possible
==25862== 
==25862== For lists of detected and suppressed errors, rerun with: -s
==25862== ERROR SUMMARY: 9 errors from 4 contexts (suppressed: 0 from 0)
```

</details>

cosmo0920 commented on plugins/out_s3/s3.c:
```diff
@@ -40,6 +40,10 @@
+#include "fluent-bit/flb_output.h"
+#include "fluent-bit/flb_sds.h"
+#include "fluent-bit/flb_uri.h"
+#include "mk_core/mk_list.h"
```
We shouldn't use `""` type of include directives for core sources and bundled sources.
It could be easily confusing to read.

```c
#include <fluent-bit/flb_output.h>
#include <fluent-bit/flb_sds.h>
#include <fluent-bit/flb_uri.h>
```

Needed to use the above instead.

Plus, we don't need to include `#include <mk_core/mk_list.h>` here.
This header is already included with plugin related headers.

dmytro-sirenko replied:
done

cosmo0920 commented on plugins/out_s3/s3.h:
```diff
@@ -20,6 +20,7 @@
 #ifndef FLB_OUT_S3
 #define FLB_OUT_S3
 
+#include "mk_core/mk_list.h"
```
We don't need to include mk_core/mk_list.h here.
It is already included in plugin related headers.

dmytro-sirenko replied:
Thanks, `clangd` included this automatically. I've cleaned it up.

dmytro-sirenko committed:
stylistic changes

Signed-off-by: Dmytro Sirenko <dmytro.sirenko@mongodb.com>

dmytro-sirenko commented on plugins/out_s3/s3.c:
```diff
@@ -583,9 +604,56 @@ static void s3_context_destroy(struct flb_s3 *ctx)
+            flb_sds_destroy(val_sds); val_sds = NULL;
+        } else {
+            if (NULL == flb_sds_printf(&url_sds, "&%s=%s", key_sds, val_sds))
+                goto error;
```
done

dmytro-sirenko commented on plugins/out_s3/s3.c:
```diff
@@ -583,9 +604,60 @@ static void s3_context_destroy(struct flb_s3 *ctx)
+                goto error;
+        }
+        if (NULL == flb_sds_printf(&url_sds, "%s=%s", key_sds, val_sds))
+            goto error;
```
fixed

dmytro-sirenko commented:
@cosmo0920 could you take a look?

runderwo commented:
@cosmo0920 @edsiper Can anyone review this?

cosmo0920 commented:
Oh, sorry for being late but we need to resolve conflicts before entering reviewing process.

dmytro-sirenko commented:
The changes are rebased and ready for review.

runderwo commented:
@cosmo0920 Can you re-check before the branch gets out of date please?

runderwo commented:
@cosmo0920 @edsiper Can someone re-review?  We are experiencing a repeating cycle of the branch going out of date by the time someone looks at it, and then review being blocked on updating the branch.
